### PR TITLE
retryWhen operator, tests, demo

### PR DESCRIPTION
**Pull request by mbonneau**

### CODE REVIEW

File: lib/Rx/Operator/RetryWhenOperator.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    public function __invoke(
+        ObservableInterface $observable,
+        ObserverInterface $observer,
+        SchedulerInterface $scheduler = null
+    ) {
+        $errors           = new Subject();
+        $when             = call_user_func($this->notificationHandler, $errors->asObservable());
+        $disposable       = new CompositeDisposable();
+        $sourceDisposable = new EmptyDisposable();
+        $innerCompleted   = false;
```
Comment (davidwdan) [inline lib/Rx/Operator/RetryWhenOperator.php:38]: Is there any reason why these can't be private properties?

File: test/Rx/Functional/Operator/RetryWhenTest.php [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+        $this->assertMessages([
+            onNext(210, 1),
+            onNext(220, 2)
+        ], $results->getMessages());
+
+        $this->assertSubscriptions([
+            subscribe(200, 230)
+        ], $xs->getSubscriptions());
+    }
```
Comment (davidwdan) [inline test/Rx/Functional/Operator/RetryWhenTest.php:270]: Should we have some dispose tests?

Comment (mbonneau) [inline test/Rx/Functional/Operator/RetryWhenTest.php:270]: I used the tests from the RxJS - I think there could be a few more cases added. I will work on that when I have a chance.

File: lib/Rx/Operator/RetryWhenOperator.php (hunk 1 above)
Comment (asm89) [inline lib/Rx/Operator/RetryWhenOperator.php:38]: We shouldn't keep state in the operators themselves I think. Nothing is stopping you from doing:

```
$myOperator = new MyOperator();
$o1->lift($myOperator)->subscribe(/* ... */);
$o2->lift($myOperator)->subscribe(/* ... */);
```

That would get messy?

Comment (asm89) [inline lib/Rx/Operator/RetryWhenOperator.php:38]: Nevermind. See https://github.com/ReactiveX/RxPHP/pull/26#discussion_r52835458.

**Comment by mbonneau** [comment]
I added a few more tests and caught a few minor issues.

File: lib/Rx/Operator/RetryWhenOperator.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                [$observer, 'onCompleted']
+            ), $scheduler);
+
+            $disposable->add($sourceDisposable);
+        };
+
+        $whenDisposable = $when->subscribe(new CallbackObserver(
+            function ($x) use ($subscribeToSource, &$sourceError) {
+                if ($sourceError) {
+                    $sourceError = false;
```
Comment (asm89) [inline lib/Rx/Operator/RetryWhenOperator.php:88]: nit: this already happens in `$subscribeToSource()`?

**Pull request merged**